snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -100,50 +103,56 @@ def submit(self, message: Message[Tick]) -> None:
             self.__offset_low_watermark = message.offset
 
     def __should_commit(self, message: Message[Tick]) -> bool:
-        return (
-            self.__offset_low_watermark is None
-            or message.offset > self.__offset_low_watermark
-        ) and (
-            self.__offset_high_watermark is not None
-            and message.offset <= self.__offset_high_watermark
-        )
+        if self.__offset_high_watermark is None:
+            return False
+
+        if self.__offset_low_watermark is None:
+            return True
+
+        return self.__offset_high_watermark > self.__offset_low_watermark
 
     def __update_offset_high_watermark(self, message: Message[Tick]) -> None:
         assert message.partition.index == 0, "Commit log cannot be partitioned"
         tick_partition = message.payload.partition
         assert tick_partition is not None
         self.__latest_messages_by_partition[tick_partition] = message
 
-        slowest = message
-        fastest = message
+        # Keep track of the slowest/fastest partition based on the tick's lower timestamp.
+        # This is only used for recording the partition lag metric.
+        slowest = message.payload.timestamps.lower
+        fastest = message.payload.timestamps.lower
+
+        # Keep track of the earliest partition based on its offset in the commit log topic.
+        # This is used to determine the offset that we can safely commit.
+        earliest = message.offset
+
         for partition_message in self.__latest_messages_by_partition.values():
             if partition_message is None:
                 return
 
             partition_timestamp = partition_message.payload.timestamps.lower
 
-            if partition_timestamp < slowest.payload.timestamps.lower:
-                slowest = partition_message
+            if partition_timestamp < slowest:
+                slowest = partition_timestamp
 
-            if partition_timestamp > fastest.payload.timestamps.lower:
-                fastest = partition_message
+            if partition_timestamp > fastest:
+                fastest = partition_timestamp
```

Comment:
Do you actually need to keep track of `slowest` and `fastest` for anything other than logging the lag ?

Reply:
No we don't, that's noted in a comment above where the variables are declared.